System UI Font Extraction should use CTFontGetUIFontType as opposed to the "." internal postfix name

gavin-apple opened the pull request:
#### 3b5afeaaf5da200e9becf2136000d85c63eafcad
<pre>
System UI Font Extraction should use CTFontGetUIFontType as opposed to the &quot;.&quot; internal postfix name
<a href="https://rdar.apple.com/164085378">rdar://164085378</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302004">https://bugs.webkit.org/show_bug.cgi?id=302004</a>

Reviewed by Brent Fulgham and Vitor Roriz.

Ports resolution of System fonts to us the SystemUI font name as
opposed to the internal &quot;.&quot; postfix name. We introduce a new
SystemUIFont InstalledFont type in order to handle serialization
in this manner.

* Source/WebCore/Configurations/AllowedSPI.toml:
* Source/WebCore/PAL/pal/spi/cf/CoreTextSPI.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::toSerializableInstalledFont const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::InstalledFont::SystemUIFont::toCTFont const):
(WebCore::InstalledFont::PostScriptFont::toCTFont const):
(WebCore::InstalledFont::toCTFont const):
(WebCore::InstalledFont::toFont const):
* Source/WebKit/Shared/WebCoreFont.serialization.in:

Canonical link: <a href="https://commits.webkit.org/304099@main">https://commits.webkit.org/304099@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff1893cfb69d9e192cee74d783cf404c6a85fc47

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134253 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6765 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45475 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141834 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86304 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a841bfd2-2916-4213-9e94-5f0ea8ad7832) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136123 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7354 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6627 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102659 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69931 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8a1cb570-ff4c-47b8-a63b-37d361c951df) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137200 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5131 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120335 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83452 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/49d7554d-a1f7-453f-b05d-c29e0be60ac1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5007 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2608 "Passed tests") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114241 "Found 2 new API test failures: TestWebKitAPI.IndexedDB.IndexedDBSuspendImminently, TestWebKitAPI.ProcessSwap.PageOverlayLayerPersistence (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38473 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144480 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6435 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39053 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111043 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6518 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5417 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111355 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28285 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4835 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116606 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60224 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6487 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34818 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6326 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70051 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6565 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6438 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz reviewed (commented):
Could we split this into 2 PRs? It is a bit hard to follow the 2 different things we are trying to do here.

vitorroriz commented on Source/WebCore/PAL/pal/spi/cf/CoreTextSPI.h:
This fails style checker:

`CTFontUIFontType CTFontGetUIFontType(CTFontRef);`

vitorroriz commented on Source/WebKit/Shared/WebCoreFont.serialization.in:
```diff
@@ -167,14 +167,21 @@ header: <WebCore/FontPlatformData.h>
     bool syntheticOblique;
 }
 
+using WebCore::SystemUIFontType = uint32_t;
```
isn't this already defined at [FontPlatformData.h‎](https://github.com/WebKit/WebKit/pull/53449/files#diff-e64200f26c5f275d1747098c5f94fcad1c936d37a698e312bdd1215676a58b0d)?

gavin-apple replied:
We need it declared in the .serialization.in files too so the type information is exposed through the IPCTestingAPI. As an example (I don't have a local build of this PR) we have:

```
for using_statement in using_statements:
        if using_statement.condition is not None:
            result.append(f'#if {using_statement.condition}')
        result.append(f'        {{ "{using_statement.name}"_s, {{')
        result.append(f'        {{')
        for line_number in range(len(using_statement.alias_lines)):
            alias_line = using_statement.alias_lines[line_number]
            if '#' in alias_line:
                result.append(f'{alias_line.strip()}')
            else:
                underscore_s_after_last_line = '_s' if line_number is len(using_statement.alias_lines) - 1 else ''
                extra_space_after_comma = ' ' if alias_line.endswith(',') else ''
                result.append(f'            "{alias_line.strip()}{extra_space_after_comma}"{underscore_s_after_last_line}')
        result.append(f'            , "alias"_s }}')
        result.append('        } },')
        if using_statement.condition is not None:
            result.append('#endif')
```

Generates:

```
{ "CFTimeInterval"_s, {
        {
            "double"_s
            , "alias"_s }
        } },
        { "CGGlyph"_s, {
        {
            "uint16_t"_s
            , "alias"_s }
        } },
#if ENABLE(WRITING_TOOLS)
        { "WebCore::WritingTools::TextSuggestion::ID"_s, {
        {
            "WTF::UUID"_s
            , "alias"_s }
        } },
#endif
#if ENABLE(WRITING_TOOLS)
        { "WebCore::WritingTools::Session::ID"_s, {
        {
            "WTF::UUID"_s
            , "alias"_s }
        } },
#endif
#if ENABLE(MEDIA_CONTROLS_CONTEXT_MENUS) && USE(UICONTEXTMENU)
        { "WebCore::MediaControlsContextMenuItem::ID"_s, {
        {
            "uint64_t"_s
            , "alias"_s }
        } },
#endif
```

from:

```
using CFTimeInterval = double
using CGGlyph = uint16_t

#if ENABLE(WRITING_TOOLS)
using WebCore::WritingTools::TextSuggestion::ID = WTF::UUID
using WebCore::WritingTools::Session::ID = WTF::UUID
#endif

#if ENABLE(MEDIA_CONTROLS_CONTEXT_MENUS) && USE(UICONTEXTMENU)
using WebCore::MediaControlsContextMenuItem::ID = uint64_t;
#endif
```

This then gets exposed to JavaScript through:

```
IPC.serializedTypeInfo["WebCore::WritingTools::TextSuggestion::ID"]
```
(and similarly for other clients that use the info).

You can see usages in CoreIPC.js for how it's then useful for Testing/Fuzzing/etc.

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -1029,6 +1029,16 @@ std::optional<InstalledFont> Font::toSerializableInstalledFont() const
         platformData().syntheticOblique()
     };
 
+    if ((SystemUIFontType)CTFontGetUIFontType(ctFont.get()) != SystemUIFontTypeNone) {
```
Should we use `static_cast` here?

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -1029,6 +1029,16 @@ std::optional<InstalledFont> Font::toSerializableInstalledFont() const
+    if ((SystemUIFontType)CTFontGetUIFontType(ctFont.get()) != SystemUIFontTypeNone) {
+        return InstalledFont {
+            InstalledFont::SystemUIFont {
+                (SystemUIFontType)CTFontGetUIFontType(ctFont.get()),
```
I guess we should avoid calling CT twice for the same input, let's hold into the result of `CTFontGetUIFontType(ctFont.get())`

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -524,40 +524,32 @@ RetainPtr<CFTypeRef> FontPlatformOpticalSize::toCF() const
-        matchedFont = adoptCF(CTFontCreateWithFontDescriptor(matchedDescriptor.get(), pointSize, nullptr));
-        if (String(adoptCF(CTFontCopyPostScriptName(matchedFont.get())).get()) == postScriptName)
-            return matchedFont;
+    return adoptCF(CTFontCreateUIFontForLanguage((CTFontUIFontType)systemUIFontType, pointSize, language.createCFString().get()));
```
another `c-style` cast here

brentfulgham commented on Source/WebCore/Configurations/AllowedSPI.toml:
```diff
@@ -71,5 +71,5 @@ selectors = [
-    "CTFontDescriptorCreateMatchingFontDescriptorsWithOptions",
-]
\ No newline at end of file
+    "CTFontGetUIFontType",
```
I confirmed the request/cleanup radars now reflect the `CTFontGetUIFontType` SPI -> API request, and that this is on track.

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -542,40 +542,32 @@ RetainPtr<CFTypeRef> FontPlatformOpticalSize::toCF() const
-        if (String(adoptCF(CTFontCopyPostScriptName(matchedFont.get())).get()) == postScriptName)
-            return matchedFont;
+    return adoptCF(CTFontCreateUIFontForLanguage(static_cast<CTFontUIFontType>(systemUIFontType), pointSize, language.createCFString().get()));
+}
```
Nit: This diff would be easier to read if you had added the new `SystemUIFont::toCTFont` after the existing `PostScriptFont::toCTFont` method!

brentfulgham commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -171,6 +171,9 @@ struct FontPlatformDataAttributes {
 #if USE(CORE_TEXT)
 
+using SystemUIFontType = uint32_t;
+#define SystemUIFontTypeNone UINT32_MAX
```
Nit: Ideally it seems like these should be concepts exposed by CoreText somehow, but I guess the concept of `SystemUIFont` is a distinction we make in WebKit, as opposed to CoreText just dealing with fonts of all flavors.

brentfulgham reviewed (approved):
I had a few minor comments for future work (or requests for clarification with the CoreText folks), but nothing to address for this patch.

r=me

gavin-apple committed:
System UI Font Extraction should use CTFontGetUIFontType as opposed to the "." internal postfix name
rdar://164085378
https://bugs.webkit.org/show_bug.cgi?id=302004

Reviewed by Brent Fulgham and Vitor Roriz.

Ports resolution of System fonts to us the SystemUI font name as
opposed to the internal "." postfix name. We introduce a new
SystemUIFont InstalledFont type in order to handle serialization
in this manner.

* Source/WebCore/Configurations/AllowedSPI.toml:
* Source/WebCore/PAL/pal/spi/cf/CoreTextSPI.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::toSerializableInstalledFont const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::InstalledFont::SystemUIFont::toCTFont const):
(WebCore::InstalledFont::PostScriptFont::toCTFont const):
(WebCore::InstalledFont::toCTFont const):
(WebCore::InstalledFont::toFont const):
* Source/WebKit/Shared/WebCoreFont.serialization.in:

Canonical link: https://commits.webkit.org/304099@main
